buildSrc/src/main/kotlin/org/gradle/modules/ClasspathManifestPatcher.kt

```diff
@@ -0,0 +1,96 @@
+package org.gradle.modules
+
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.ResolvedDependency
+import org.gradle.util.CollectionUtils
+import org.gradle.util.GUtil
+
+import org.gradle.kotlin.dsl.*
+
+import org.objectweb.asm.ClassReader
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.commons.ClassRemapper
+import org.objectweb.asm.commons.Remapper
+
+import java.io.File
+
+
+// Map of internal APIs that have been renamed since the last release
+val REMAPPINGS = mapOf("org/gradle/plugin/use/internal/PluginRequests" to "org/gradle/plugin/management/internal/PluginRequests")
+
+open class ClasspathManifestPatcher(val project: Project, val temporaryDir: File, val runtime: Configuration, val moduleNames: Set<String>) {
+    fun writePatchedFilesTo(outputDir: File) {
+        resolveExternalModuleJars().forEach {
+            val originalFile = mainArtifactFileOf(it)
+            val patchedFile = File(outputDir, originalFile.name)
+            val unpackDir = unpack(originalFile)
+            patchManifestOf(it, unpackDir)
+            patchClassFilesIn(unpackDir)
+            pack(unpackDir, patchedFile)
+        }
+    }
+
+    /**
+     * Resolves each external module against the runtime configuration.
+     */
+    private fun resolveExternalModuleJars(): Collection<ResolvedDependency> {
+        return runtime.resolvedConfiguration.firstLevelModuleDependencies
+            .filter { it.moduleName in moduleNames }
+    }
+
+    private fun patchManifestOf(module: ResolvedDependency, unpackDir: File) {
+        val classpathManifestFile = File(unpackDir, "${module.moduleName}-classpath.properties")
+        val classpathManifest = GUtil.loadProperties(classpathManifestFile)
+        classpathManifest["runtime"] = runtimeManifestOf(module)
+        classpathManifestFile.writeText(classpathManifest.map { "${it.key}=${it.value}" }.joinToString { "\n" })
```

Comment:
:x: That `map ... joinToString ...` sequence is not preserving behaviour, a correct and more efficient translation would be `classpathManifest.joinToString(separator = "\n") { "${it.key}=${it.value}" }`.